### PR TITLE
Adds retry logic to integration tests

### PR DESCRIPTION
**Pull request by mpmcroy**
Added retry logic to rerun failed tests up to 5 times. This should help with flakey tests.

It took me much longer that it should have to get this to work both locally and on Travis 😄

### CODE REVIEW

File: src/shadowbox/integration_test/test.sh [hunk 1]
```diff
@@ -81,6 +81,29 @@ function cleanup() {
...
+function test_with_retries() {
+  local test_command="${1}"             # Test command to run
+  local expected_result_code="${2}"     # Result code to expected for test command
+  local failure_message="${3}"          # Message to display on failure
+  local max_attempts="${4:-5}"          # Maximum number of attempts
+
+  local RETRY_DELAY_SECONDS=5
+  local attempt_count=1
+
+  until (${test_command} > /dev/null || (($? == ${expected_result_code}))); do
```
Comment (fortuna) [inline src/shadowbox/integration_test/test.sh:95]: Use a for loop instead: https://linuxize.com/post/bash-for-loop/#the-c-style-bash-for-loop

File: src/shadowbox/integration_test/test.sh [hunk 2]
```diff
@@ -109,8 +132,8 @@ function cleanup() {
     fail "Client should not have access to target IP" || (($? == 28))
 
   # Exit code 6 for "Could not resolve host".
-  docker exec $CLIENT_CONTAINER curl --silent --connect-timeout 1 http://target > /dev/null && \
-    fail "Client should not have access to target host" || (($? == 6))
+  test_with_retries "docker exec $CLIENT_CONTAINER curl --silent --connect-timeout 1 http://target" \
```
Comment (fortuna) [inline src/shadowbox/integration_test/test.sh:135]: It's a bit harder to read this now. There are two failure conditions and test_with_retries is doing too much.

What do you think about creating a function that runs the previous test, then pass it to a retry function?
Example:
```bash
function client_does_not_have_access_to_target_host() {
  docker exec $CLIENT_CONTAINER curl --silent --connect-timeout 1 http://target > /dev/null && fail "Client should not have access to target host" || (($? == 6))
}
...
test_with_retries client_does_not_have_access_to_target_host
...
test_with_retries deleted_access_key_is_inactive
```

File: src/shadowbox/integration_test/test.sh (hunk 1 above)
Comment (mpmcroy) [reply src/shadowbox/integration_test/test.sh]: Done

File: src/shadowbox/integration_test/test.sh (hunk 2 above)
Comment (mpmcroy) [reply src/shadowbox/integration_test/test.sh]: That's a great suggestion. Done.

**Comment by mpmcroy** [comment]
> Thanks for tackling this. I'm still curious why the test is flaky. Maybe we should surface the curl error message somehow.

I did a little analysis to see which tests are flaky and I shared the doc with you. 
* `client_does_not_have_access_to_target_host` is failing with curl error code 28 (operation timeout) when we expect it to fail with error code 6 (could not resolve host). We could increase the timeout, but the retry should work.
* `deleted_access_key_is_inactive` is failing with curl error code code 52 (got nothing). We'll have to look into that a bit more if retrying doesn't work.

File: src/shadowbox/integration_test/test.sh [hunk 3]
```diff
@@ -81,6 +81,34 @@ function cleanup() {
...
+# Runs a test command and checks for expected result code and retries on failure.
+function test_with_retries() {
+  local test_function=${1}
+  local max_attempts=${2:-5}
+
+  for ((i=1; i<=${max_attempts}; i++))
+  do
+    output=$(${test_function})
+    result_code=$?
+    if [ $result_code -eq 0 ]; then
```
Comment (fortuna) [inline src/shadowbox/integration_test/test.sh:93]: Use `[[ $result_code == 0 ]]` instead: https://google.github.io/styleguide/shellguide.html#s6.3-tests

**Comment by fortuna** [comment]
It seems the integration test is failing, and somewhere else not covered by the retries. Maybe it's a legit breakage?

**Comment by mpmcroy** [comment]
My assumption was that the tests were flaky and that retrying them would help, but that might not be the case :/

Let me hold off on this for a minute and add retry to all tests. I want to run this on Travis a few times to see if retrying will actually help address the issue.

File: src/shadowbox/integration_test/test.sh [hunk 4]
```diff
@@ -88,12 +88,13 @@ function test_with_retries() {
 
   for ((i=1; i<=${max_attempts}; i++))
   do
-    output=$(${test_function})
-    result_code=$?
-    if [[ $result_code -eq 0 ]]; then
+    # The script will exit if the below test returns with a non-zero status code
+    # (due to set -e), so we capture the result code separately.
+    $(${test_function}) && result_code=$? || result_code=$?
```
Comment (fortuna) [inline src/shadowbox/integration_test/test.sh:93]: Remove the `$()`. No need for that, the test_function call directly should work.

Comment (fortuna) [reply src/shadowbox/integration_test/test.sh:93]: I think you can also do this, which is clearer:
```bash
if ${test_function}; then
  return
else 
  result_code=$?
fi
```

**Pull request closed without merging**